MetabolicAtlas/MetabolicAtlas
# feat: case insensitive search/filtering for gotenzymes

**Related issue(s) and PR(s)**  
This PR closes #1110  and relates to #1112 .

Searching on the GotEnzymes landing page is now case insensitive. Filtering columns of the enzymes table is also case insensitive.


**Type of change**  


 
**List of changes made**  

- use lower case for the db queries when searching / filtering
- add indixes for lower cased fields to the psql db
- add api tests

**Testing**  

To create the new indexes, either reconstructing the psql db:
`ma-exec pg psql -f scripts/init.sql -U postgres`
or by do this :
` psql -h localhost -p 5432 -U postgres`
(use password from `env-local.env`) and then run:
```
  create index lower_cased_genes on genes ((lower(kegg)));
  create index lower_genes on enzymes ((lower(gene))) ;
  create index lower_domain on enzymes (lower(domain));
  create index lower_organism on enzymes (lower(organism));
  create index lower_compound on enzymes (lower(compound));
  create index lower_reaction on enzymes (lower(reaction_id));
```
This takes some minutes.

- Go to `/gotenzymes` and verify that case is not important when searching, for all catergories (genes, domains, organisms, reactions, compouns). For example, search for `atp6` and make sure that 2 genes are listed in the results
- Visit a details page, eg `/gotenzymes/reaction/R01246` and try filtering by gene `acav_1835`. The resulting table should list `Acav_1835`.
- Also try `/gotenzymes/domain/B` (this is a bit slow, see *Further comments* below). Filter by `b1468` and make sure you see both `b1468` and `B1468`.

**Further comments**  

For most cases, the performance is good on my local machine. The slowest page I found is `/gotenzymes/domain/B`. It is also the one containing the most rows. What actually makes it slow is the total count, which the db for some reason decides to handle quite differently from before (`Parallel Bitmap Heap Scan` instead of `Parallel Index Only Scan`). If the speed is a problem, here are some thoughts:
- do some more testing / researching
- have a separate call for getting the total count so that the first page can be shown as soon as possible
- have a separate call for creating the initial enzymes table for a details page, that uses exact match search. We know that the initial filter (eg `domain = 'B'`) will have the correct case.
- the filtering of the columns are subject to change in #1056.

**Definition of Done checklist**
